framework/src/main/java/org/tron/core/net/peer/PeerConnection.java

```diff
@@ -304,6 +329,117 @@ public synchronized boolean checkAndPutAdvInvRequest(Item key, Long value) {
     return true;
   }
 
+  // if peer is in adv status and no block received and sent between us for too long, and we
+  // don't receive block inventory also, then the peer is not active.
+  public boolean isNotActiveTooLong() {
+    return System.currentTimeMillis() - getLatestTime()
+        > resilienceConfig.getPeerNotActiveThreshold() * 1000L;
+  }
+
+  private long getLatestTime() {
+    List<Long> times = Arrays.asList(
+        channel.getLastActiveTime(), //last block time
+        getMaliciousFeature().getAdvStartTime(), //adv begin time
+        getMaliciousFeature().getLastRecBlockInvTime()); //inventory
+    return Collections.max(times);
+  }
+
+  @Getter
+  public class MaliciousFeature {
+
+    @Setter
+    private long advStartTime = System.currentTimeMillis();
+    @Setter
+    private long stopBlockInvStartTime = -1;
+    @Setter
+    private long stopBlockInvEndTime = -1;
+    @Setter
+    private long lastRecBlockInvTime = -1;
+    //if testStopInv=true, we use feature 4, else use feature 3. We an only use one of them.
+    private final boolean testStopInv;
+
+    //four features
+    private long badSyncBlockChainTime = -1; //feature 1
+    private long badChainInventoryTime = -1; //feature 2
+    private long zombieBeginTime = -1;       //feature 3
+    private long zombieBeginTime2 = -1;      //feature 4
+
+    public MaliciousFeature() {
+      testStopInv = Args.getInstance().getResilienceConfig().isTestStopInv();
+    }
+
+    //it can only be set from -1 to positive
+    public void updateBadFeature1() {
```

Comment:
The design is unreasonable. All features should be moved to the ResilienceService you defined for implementation. This class only needs to define fields and set field values.